python/cudf/cudf/comm/gpuarrow.py

```diff
@@ -1,3 +1,183 @@
+        return self._series.data.mem
+    @property
+    def null_raw(self):
+        "Accessor for the null buffer as a device array"
+        return self._series.nullmask.mem
+    def make_series(self):
+        """Make a Series object out of this node
+        """
+        return self._series.copy(deep=False)
+    def _make_dictionary_series(self):
+        """Make a dictionary-encoded series from this node
+        """
+        assert self.is_dictionary
+        return self._series.copy(deep=False)
+
+def gpu_view_as(buf, dtype, shape=None, strides=None):
+    ptr = numba.cuda.cudadrv.driver.device_pointer(buf.to_numba())
+    return rmm.device_array_from_ptr(ptr, buf.size // dtype.itemsize, dtype=dtype)
+
+def make_device_arrays(array):
+
+    buffers = array.buffers()
+    dtypes = [np.dtype(np.int8), None, None]
+
+    if pa.types.is_list(array.type):
+        dtypes[1] = np.dtype(np.int32)
+    elif pa.types.is_string(array.type) or pa.types.is_binary(array.type):
+        dtypes[2] = np.dtype(np.int8)
+        dtypes[1] = np.dtype(np.int32)
+    elif not pa.types.is_dictionary(array.type):
+        dtypes[1] = np.dtype(array.type.to_pandas_dtype())
+    else:
+        dtypes[1] = np.dtype(array.type.index_type.to_pandas_dtype())
+
+    for i in range(len(buffers)):
+        buffers[i] = None if buffers[i] is None \
+            else gpu_view_as(CudaBuffer.from_buffer(buffers[i]), dtypes[i])
+
+    return buffers
+
+def array_to_series(array):
+
+    if isinstance(array, pa.ChunkedArray):
+        return Series._concat([array_to_series(chunk) for chunk in array.chunks])
+    if isinstance(array, pa.Column):
+        return Series._concat([array_to_series(chunk) for chunk in array.data.chunks])
+
+    dtype = None
+    array_len = len(array)
+    null_count = array.null_count
+    buffers = make_device_arrays(array)
+    mask, data = buffers[0], buffers[1]
+
+    if pa.types.is_dictionary(array.type):
+        from cudf.dataframe import CategoricalColumn
+        dtype = 'categorical'
+        codes = array_to_series(array.indices)
+        categories = array_to_series(array.dictionary)
+        data = CategoricalColumn(data=codes.data,
+                                 mask=mask,
+                                 null_count=null_count,
+                                 categories=categories,
+                                 ordered=array.type.ordered)
+    elif pa.types.is_string(array.type):
+        import nvstrings
+        offs, data = buffers[1], buffers[2]
+        offs = offs[array.offset : array.offset + array_len + 1]
+        data = np.empty(0, dtype=np.dtype(np.int8)) if data is None else data
+        mask = None if mask is None else mask.copy_to_host().view(np.dtype(np.int8))
+        data = nvstrings.from_offsets(data.copy_to_host(),
+                                      offs.copy_to_host(),
+                                      array_len, nbuf=mask, ncount=null_count)
```

Comment:
@davidwendt here's the spot where I was having trouble. data, mask, and offsets are all device buffers now, but couldn't get past this call without copying them to the host here. I tried setting `bdevmem=True`, but no luck.

Does `from_offsets()` want to be passed the numba `device_pointer` as a uint rather than the python DeviceNDArrays? Thanks.

Reply:
I think it should be able to handle DeviceNDArrays but I would have to try it out. It is suppose to work with python buffer interface objects. Regardless here is my python test script for from_offsets that may help for now:
https://github.com/rapidsai/custrings/blob/ae3e81b9ed83eff6e01d8977fab086f9d6e209f2/python/tests/test_offsets.py